File: v4/rust/iprange-livedb/src/validation.rs
//! Explicit bounded full-file validation.

// A validation failure returns its factual progress and cleanup authority inline.
#![allow(clippy::result_large_err)]

mod bitmap;
mod blob;
mod catalog;
mod context;
mod membership;
mod membership_table;
mod metadata;
mod page;
mod range;
mod retirement;
pub(crate) mod source;
mod structure;
mod structure_table;
mod tree;
mod types;

use std::path::Path;

use crate::bootstrap::{BootstrapError, MetaProblem, OpenMode};
use crate::cancellation::CancellationToken;
use crate::contract::{MetaV4, PAGE_SIZE};
use crate::database_file;
use crate::error::{Error, Result};
use crate::mapping::Mapping;
use crate::publication::{CleanupArtifacts, CoordinationCleanup};
use crate::recovery::RecoverySourceCleanupGuard;

use source::{ImmutableSource, LiveBootstrapSource, LiveOpened, LiveSource};

pub use types::{
    LocalFileIdentity, PhysicalByteInterval, ValidatedGeneration, ValidationAddressFence,
    ValidationBudget, ValidationFailure, ValidationFinding, ValidationMode, ValidationObject,
    ValidationProgress, ValidationReason, ValidationResult, ValidationSink, ValidationSinkControl,
};

/// Validate one explicitly selected source mode without changing the source.
/// Factual availability facts for the version-matched fault worker
/// used by validation and recovery (spec `system.describe.fault_worker`).
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct WorkerAvailability {
    /// True when a candidate worker executable exists beside the binary.
    pub available: bool,
    /// Worker control-protocol version understood by this build.
    pub protocol: &'static str,
}

/// Reports fault-worker availability for this installation without
/// spawning the worker.
pub fn worker_availability() -> WorkerAvailability {
    crate::worker::availability()
}

pub fn validate<S: ValidationSink>(
    path: impl AsRef<Path>,
    mode: ValidationMode,
    budget: &ValidationBudget,
    cancellation: &CancellationToken,
    sink: &mut S,
) -> std::result::Result<ValidationResult, ValidationFailure> {
    if let Err(cause) = preflight(&mode, budget, cancellation) {
        return Err(failure(cause, ValidationProgress::new()));
    }
    crate::worker::validate(path.as_ref(), &mode, budget, cancellation, sink)
}

pub(crate) fn validate_local<S: ValidationSink>(
    path: &Path,
    mode: ValidationMode,
    budget: &ValidationBudget,
    cancellation: &CancellationToken,
    sink: &mut S,
) -> std::result::Result<ValidationResult, ValidationFailure> {
    if let Err(cause) = preflight(&mode, budget, cancellation) {
        return Err(failure(cause, ValidationProgress::new()));
    }
    match mode {
        ValidationMode::ImmutableCurrent => {
            if budget.max_open_files < 1 {
                return Err(failure(
                    Error::BudgetExceeded("immutable validation open files"),
                    ValidationProgress::new(),
                ));
            }
            validate_immutable(path, budget, cancellation, sink)
        }
        ValidationMode::LiveCurrent => {
            if budget.max_open_files < 2 {
                return Err(failure(
                    Error::BudgetExceeded("live validation open files"),
                    ValidationProgress::new(),
                ));
            }
            validate_live(path, budget, cancellation, sink)
        }
        ValidationMode::OfflineCandidate(candidate) => {
            validate_offline(path, &candidate, budget, cancellation, sink)
        }
    }
}

fn preflight(
    mode: &ValidationMode,
    budget: &ValidationBudget,
    cancellation: &CancellationToken,
) -> Result<()> {
    if matches!(mode, &ValidationMode::LiveCurrent) {
        crate::live_lock::require_live_supported()?;
    }
    budget.validate()?;
    cancellation.check()
}

fn validate_offline<S: ValidationSink>(
    path: &Path,
    candidate: &crate::recovery::RecoveryCandidate,
    budget: &ValidationBudget,
    cancellation: &CancellationToken,
    sink: &mut S,
) -> std::result::Result<ValidationResult, ValidationFailure> {
    let source = crate::recovery::inspection::OfflineSource::open(path, cancellation)
        .map_err(|cause| failure(cause, ValidationProgress::new()))?;
    let identity = source.public_identity();
    let classified = crate::recovery::inspection::read_classified(&source.file, cancellation)
        .map_err(|cause| failure(cause, ValidationProgress::new()))?;
    if identity != candidate.source_identity {
        return Err(failure(
            Error::RecoveryCandidateChanged,
            ValidationProgress::new(),
        ));
    }
    let meta = classified
        .selected_meta(candidate)
        .ok_or_else(|| failure(Error::RecoveryCandidateChanged, ValidationProgress::new()))?;
    source
        .require_available(meta.database_id)
        .map_err(|cause| failure(cause, ValidationProgress::new()))?;
    let mapping = validation_mapping(&source.file, meta)
        .map_err(|cause| failure(cause, ValidationProgress::new()))?;
    let mut context = context::Context::new(&mapping, meta, budget, cancellation, sink)
        .map_err(|cause| failure(cause, ValidationProgress::new()))?;
    let scan = crate::worker::probe_source(&mapping, || {
        context
            .reserve_allocator_pages()
            .and_then(|()| validate_selected(&mut context))
    });
    let verification = verify_offline_candidate(&source, candidate, cancellation);
    let progress = context.finish();
    if let Err(cause) = scan {
        return Err(failure(cause, progress));
    }
    if let Err(cause) = verification {
        return Err(failure(cause, progress));
    }
    Ok(ValidationResult {
        valid: progress.finding_count == 0,
        file_identity: identity,
        generation: Some(generation(meta)),
        progress,
    })
}

fn verify_offline_candidate(
    source: &crate::recovery::inspection::OfflineSource,
    candidate: &crate::recovery::RecoveryCandidate,
    cancellation: &CancellationToken,
) -> Result<()> {
    source.verify().map_err(candidate_identity_error)?;
    let classified = crate::recovery::inspection::read_classified(&source.file, cancellation)?;
    if classified.selected_meta(candidate).is_none() {
        return Err(Error::RecoveryCandidateChanged);
    }
    source.verify().map_err(candidate_identity_error)
}

fn candidate_identity_error(cause: Error) -> Error {
    match cause {
        Error::WrongMode(_) => Error::RecoveryCandidateChanged,
        Error::Io(error) if error.kind() == std::io::ErrorKind::NotFound => {
            Error::RecoveryCandidateChanged
        }
        cause => cause,
    }
}

fn validate_live<S: ValidationSink>(
    path: &Path,
    budget: &ValidationBudget,
    cancellation: &CancellationToken,
    sink: &mut S,
) -> std::result::Result<ValidationResult, ValidationFailure> {
    match LiveSource::open(path, cancellation).map_err(|failure| {
        failure_with_guard(failure.cause, ValidationProgress::new(), failure.guard)
    })? {
        LiveOpened::Selected(source) => validate_live_selected(source, budget, cancellation, sink),
        LiveOpened::Bootstrap(source, problem) => validate_live_bootstrap(source, problem, sink),
    }
}

fn validate_live_selected<S: ValidationSink>(
    source: LiveSource,
    budget: &ValidationBudget,
    cancellation: &CancellationToken,
    sink: &mut S,
) -> std::result::Result<ValidationResult, ValidationFailure> {
    let meta = *source.meta;
    let file_identity = source.public_identity();
    let mapping = match validation_mapping(&source.file, meta) {
        Ok(mapping) => mapping,
        Err(cause) => {
            let end = source.finish(Err(cause));
            return Err(failure_with_guard(
                end.cause.expect("failed validation retains its cause"),
                ValidationProgress::new(),
                end.guard,
            ));
        }
    };
    let mut context = match context::Context::new(&mapping, meta, budget, cancellation, sink) {
        Ok(context) => context,
        Err(cause) => {
            let end = source.finish(Err(cause));
            return Err(failure_with_guard(
                end.cause.expect("failed validation retains its cause"),
                ValidationProgress::new(),
                end.guard,
            ));
        }
    };
    let scan = crate::worker::probe_source(&mapping, || {
        context
            .reserve_allocator_pages()
            .and_then(|()| validate_selected(&mut context))
    });
    let progress = context.finish();
    let scan = scan.and_then(|()| crate::worker::checkpoint_validation_progress(&progress));
    let end = source.finish(scan);
    if let Some(cause) = end.cause {
        return Err(failure_with_guard(cause, progress, end.guard));
    }
    Ok(ValidationResult {
        valid: progress.finding_count == 0,
        file_identity,
        generation: Some(generation(meta)),
        progress,
    })
}

fn validate_live_bootstrap<S: ValidationSink>(
    source: LiveBootstrapSource,
    problem: BootstrapError,
    sink: &mut S,
) -> std::result::Result<ValidationResult, ValidationFailure> {
    let file_identity = source.public_identity();
    let mut progress = ValidationProgress::new();
    let report = write_bootstrap_findings(problem, sink, &mut progress)
        .and_then(|()| progress.mark_untraversable(true));
    let end = source.finish(report);
    if let Some(cause) = end.cause {
        return Err(failure_with_guard(cause, progress, end.guard));
    }
    Ok(ValidationResult {
        valid: false,
        file_identity,
        generation: None,
        progress,
    })
}

fn validate_immutable<S: ValidationSink>(
    path: &Path,
    budget: &ValidationBudget,
    cancellation: &CancellationToken,
    sink: &mut S,
) -> std::result::Result<ValidationResult, ValidationFailure> {
    let source = ImmutableSource::open(path, cancellation)
        .map_err(|cause| failure(cause, ValidationProgress::new()))?;
    let bootstrap =
        match database_file::bootstrap_file_faultable(&source.file, OpenMode::ImmutableReader) {
            Ok(bootstrap) => bootstrap,
            Err(Error::Format(problem)) => {
                require_bound_source_available(&source)
                    .map_err(|cause| failure(cause, ValidationProgress::new()))?;
                return bootstrap_report(&source, problem, sink);
            }
            Err(cause) => return Err(failure(cause, ValidationProgress::new())),
        };
    source
        .require_available(bootstrap.meta.database_id)
        .map_err(|cause| failure(cause, ValidationProgress::new()))?;
    let mapping = validation_mapping(&source.file, bootstrap.meta)
        .map_err(|cause| failure(cause, ValidationProgress::new()))?;
    let mut context = context::Context::new(&mapping, bootstrap.meta, budget, cancellation, sink)
        .map_err(|cause| failure(cause, ValidationProgress::new()))?;
    let scan = crate::worker::probe_source(&mapping, || {
        context
            .reserve_allocator_pages()
            .and_then(|()| validate_selected(&mut context))
    });
    let verification = source.verify();
    let progress = context.finish();
    if let Err(cause) = scan {
        return Err(failure(cause, progress));
    }
    if let Err(cause) = verification {
        return Err(failure(cause, progress));
    }
    Ok(ValidationResult {
        valid: progress.finding_count == 0,
        file_identity: source.public_identity(),
        generation: Some(generation(bootstrap.meta)),
        progress,
    })
}

fn validation_mapping(file: &std::fs::File, meta: MetaV4) -> Result<Mapping> {
    let len = meta
        .page_count
        .checked_mul(PAGE_SIZE as u64)
        .ok_or(Error::ArithmeticOverflow("validation mapping length"))?;
    let mut mapping = Mapping::read_only_view(file, len)?;
    mapping.set_unreadable_pages(&crate::worker::unreadable_source_pages())?;
    Ok(mapping)
}

fn require_bound_source_available(source: &ImmutableSource) -> Result<()> {
    match source::selected_or_bound_database_id(&source.file) {
        Ok(database_id) => source.require_available(database_id),
        Err(Error::Format(_) | Error::Corrupt(_)) => Ok(()),
        Err(cause) => Err(cause),
    }
}

fn bootstrap_report<S: ValidationSink>(
    source: &ImmutableSource,
    problem: BootstrapError,
    sink: &mut S,
) -> std::result::Result<ValidationResult, ValidationFailure> {
    let mut progress = ValidationProgress::new();
    let report = write_bootstrap_findings(problem, sink, &mut progress);
    if let Err(cause) = report {
        return Err(failure(cause, progress));
    }
    if let Err(cause) = progress.mark_untraversable(true) {
        return Err(failure(cause, progress));
    }
    if let Err(cause) = source.verify() {
        return Err(failure(cause, progress));
    }
    Ok(ValidationResult {
        valid: false,
        file_identity: source.public_identity(),
        generation: None,
        progress,
    })
}

fn write_bootstrap_findings<S: ValidationSink>(
    problem: BootstrapError,
    sink: &mut S,
    progress: &mut ValidationProgress,
) -> Result<()> {
    match problem {
        BootstrapError::NoBootstrapMeta { meta0, meta1 } => {
            report_meta_problem(sink, progress, 0, meta0)
                .and_then(|()| report_meta_problem(sink, progress, 1, meta1))
        }
        BootstrapError::StaticIdentityMismatch => report_bootstrap_finding(
            sink,
            progress,
            ValidationReason::MetaStaticMismatch,
            ValidationObject::Meta,
            None,
        ),
        BootstrapError::FileTooShort
        | BootstrapError::FileUnaligned
        | BootstrapError::HostAddressability
        | BootstrapError::ImmutableLengthMismatch => report_bootstrap_finding(
            sink,
            progress,
            ValidationReason::FileGeometryInvalid,
            ValidationObject::FileGeometry,
            None,
        ),
        BootstrapError::TransactionGap
        | BootstrapError::PhysicalParity
        | BootstrapError::EqualTransactionDisagreement
        | BootstrapError::CurrentGenerationUnprovable
        | BootstrapError::UnsupportedStructure(_) => report_bootstrap_finding(
            sink,
            progress,
            ValidationReason::MetaInvalid,
            ValidationObject::Meta,
            None,
        ),
    }
}

fn report_meta_problem<S: ValidationSink>(
    sink: &mut S,
    progress: &mut ValidationProgress,
    page_number: u32,
    problem: MetaProblem,
) -> Result<()> {
    let reason = if crate::worker::source_page_unreadable(page_number) {
        ValidationReason::IoError
    } else if problem == MetaProblem::Magic {
        ValidationReason::MetaUnavailable
    } else {
        ValidationReason::MetaInvalid
    };
    report_bootstrap_finding(
        sink,
        progress,
        reason,
        ValidationObject::Meta,
        Some(page_number),
    )
}

fn report_bootstrap_finding<S: ValidationSink>(
    sink: &mut S,
    progress: &mut ValidationProgress,
    reason: ValidationReason,
    object: ValidationObject,
    page_number: Option<u32>,
) -> Result<()> {
    progress.count_finding(reason)?;
    let finding = ValidationFinding {
        sequence: progress.finding_count,
        reason,
        object,
        page_number,
        physical_bytes: page_number.map(|page| PhysicalByteInterval {
            start: u64::from(page) * crate::contract::PAGE_SIZE as u64,
            end_exclusive: (u64::from(page) + 1) * crate::contract::PAGE_SIZE as u64,
        }),
        related_page_number: None,
        address_fence: None,
    };
    crate::worker::checkpoint_validation_progress(progress)?;
    match sink.finding(&finding) {
        Ok(ValidationSinkControl::Continue) => Ok(()),
        Ok(ValidationSinkControl::Stop) => Err(Error::StoppedBySink),
        Err(cause) => Err(Error::SinkFailed(Box::new(cause))),
    }
}

pub(crate) fn failure(cause: Error, progress: ValidationProgress) -> ValidationFailure {
    failure_with_guard(cause, progress, None)
}

fn failure_with_guard(
    cause: Error,
    progress: ValidationProgress,
    source_cleanup: Option<RecoverySourceCleanupGuard>,
) -> ValidationFailure {
    let coordination_cleanup = if source_cleanup.is_some() {
        CoordinationCleanup::CleanupGuard
    } else {
        CoordinationCleanup::None
    };
    ValidationFailure {
        cause,
        progress: Box::new(progress),
        cleanup: Box::new(CleanupArtifacts::new()),
        coordination_cleanup,
        source_cleanup,
    }
}

fn validate_selected<S: ValidationSink>(context: &mut context::Context<'_, S>) -> Result<()> {
    range::validate(context)?;
    catalog::validate(context)?;
    structure::validate(context)?;
    membership::validate(context)?;
    metadata::validate(context)?;
    let free_root = context.meta.free_bitmap_root;
    let page_count = context.meta.page_count;
    bitmap::validate(context, free_root, page_count, bitmap::Kind::Free)?;
    retirement::validate(context)?;
    context.validate_partition()
}

fn generation(meta: MetaV4) -> ValidatedGeneration {
    ValidatedGeneration {
        address_family: meta.address_family,
        value_kind: meta.value_kind,
        structure_kind: meta
            .structure_kind()
            .expect("bootstrap rejects unsupported structures"),
        value_tag: meta.value_tag,
        database_id: meta.database_id,
        transaction_id: meta.txn_id,
        commit_nonce: meta.commit_nonce,
        page_count: meta.page_count,
        roots: meta.roots(),
    }
}

File: v4/go/internal/validation/validation.go
package validation

// Explicit bounded full-file validation (Rust validation.rs): the
// entry selects one source mode, preflights the budget and the
// platform, opens the locked source, bootstraps the meta pair, maps
// the committed extent read-only, and sweeps every page through the
// claims partition, streaming findings into the sink. Validation
// never modifies the source. The offline-candidate arm composes the
// shared sweep through SweepSelected; the recovery package owns the
// candidate selection and terminal (recovery/inspection.rs parity),
// because the Go mode enum cannot carry the candidate payload of the
// Rust ValidationMode::OfflineCandidate variant.

import (
	"os"

	"github.com/firehol/iprange/v4/go/internal/bootstrap"
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/live"
	"github.com/firehol/iprange/v4/go/internal/mapping"
	"github.com/firehol/iprange/v4/go/internal/publication"
)

// Validate runs one explicit validation over the selected source
// without changing it (Rust validation::validate). Exactly one of the
// result and the failure is non-nil; the failure carries the partial
// progress and the cleanup ledger.
func Validate(path string, mode ValidationMode, budget *ValidationBudget, check func() error, sink ValidationSink) (*ValidationResult, *ValidationFailure) {
	if err := preflight(mode, budget, check); err != nil {
		progress := NewProgress()
		return nil, failureOf(err, &progress)
	}
	switch mode {
	case ValidationModeImmutableCurrent:
		if budget.MaxOpenFiles < 1 {
			progress := NewProgress()
			return nil, failureOf(&format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "immutable validation open files"}, &progress)
		}
		return validateImmutable(path, budget, check, sink)
	case ValidationModeLiveCurrent:
		if budget.MaxOpenFiles < 2 {
			progress := NewProgress()
			return nil, failureOf(&format.Error{Code: format.CodeInsufficientResourceBudget, Detail: "live validation open files"}, &progress)
		}
		return validateLive(path, budget, check, sink)
	case ValidationModeOfflineCandidate:
		// The Go mode enum cannot carry the candidate payload of the
		// Rust variant; the offline-candidate arm lives in the
		// recovery package (RecoveryCandidate token + source), which
		// composes the shared sweep below.
		progress := NewProgress()
		return nil, failureOf(&format.Error{Code: format.CodeInvalidArgument, Detail: "offline-candidate validation requires the candidate token (recovery.ValidateOfflineCandidate)"}, &progress)
	default:
		progress := NewProgress()
		return nil, failureOf(&format.Error{Code: format.CodeInvalidEnum, Detail: "invalid validation mode"}, &progress)
	}
}

// Preflight checks the platform, the budget, and the cancellation
// state before any path access (Rust validation::preflight; the
// facade routing reproduces the public entry's preflight before it
// routes to the worker client).
func Preflight(mode ValidationMode, budget *ValidationBudget, check func() error) error {
	return preflight(mode, budget, check)
}

// preflight checks the platform, the budget, and the cancellation
// state before any path access (Rust validation::preflight).
func preflight(mode ValidationMode, budget *ValidationBudget, check func() error) error {
	if mode == ValidationModeLiveCurrent {
		if err := live.CheckSupported(); err != nil {
			return err
		}
	}
	if budget == nil {
		return &format.Error{Code: format.CodeInvalidArgument, Detail: "validation budget is required"}
	}
	if err := budget.validate(); err != nil {
		return err
	}
	return live.Checkpoint(check)
}

// validateImmutable runs the immutable sweep (Rust
// validate_immutable): open the locked source, bootstrap the meta
// pair, probe the allocator reserve, sweep the graph and partition,
// then re-verify the source identity under the still-held lock.
func validateImmutable(path string, budget *ValidationBudget, check func() error, sink ValidationSink) (*ValidationResult, *ValidationFailure) {
	source, err := OpenImmutableSource(path, check)
	if err != nil {
		progress := NewProgress()
		return nil, failureOf(err, &progress)
	}
	result, failure := sweepImmutable(source, path, budget, check, sink)
	return result, failure
}

// sweepImmutable runs the mapped sweep over an opened immutable
// source (Rust validate_immutable bootstrap + validation_mapping +
// reserve_allocator_pages + validate_selected + source.verify).
func sweepImmutable(source *ImmutableSource, path string, budget *ValidationBudget, check func() error, sink ValidationSink) (*ValidationResult, *ValidationFailure) {
	// The geometry is proved before any mapping exists (Rust
	// require_geometry inside bootstrap_file_faultable): a short or
	// unaligned main reports the FileGeometryInvalid finding instead
	// of failing the bootstrap mapping.
	physical := sourceFileSize(source)
	if physical < 2*format.PageSize {
		progress := NewProgress()
		return bootstrapGeometryReport(source, sink, bootstrap.Problem(&format.Error{Code: format.CodeFormatInvalid, Detail: "file smaller than two pages"}, bootstrap.ProblemFileTooShort), &progress)
	}
	if physical%format.PageSize != 0 {
		progress := NewProgress()
		return bootstrapGeometryReport(source, sink, bootstrap.Problem(&format.Error{Code: format.CodeFormatInvalid, Detail: "file size not page-aligned"}, bootstrap.ProblemFileUnaligned), &progress)
	}
	m, err := mapping.MapFile(source.FileHandle(), 2*format.PageSize, false)
	if err != nil {
		progress := NewProgress()
		return nil, failureOf(sourceCloseFold(source, err), &progress)
	}
	defer func() {
		_ = m.Close()
	}()

	p0, err := m.Page(0)
	if err != nil {
		progress := NewProgress()
		return nil, failureOf(sourceCloseFold(source, err), &progress)
	}
	p1, err := m.Page(1)
	if err != nil {
		progress := NewProgress()
		return nil, failureOf(sourceCloseFold(source, err), &progress)
	}
	res, err := bootstrap.Open(p0, p1, physical, bootstrap.ModeImmutableReader)
	if err != nil {
		if problem, ok := bootstrap.AsProblem(err); ok {
			if problem.Kind == bootstrap.ProblemNoBootstrapMeta || sourceBootstrapReportable(problem) {
				progress := NewProgress()
				if reportErr := writeBootstrapFindings(problem, sink, &progress); reportErr != nil {
					return nil, failureOf(sourceCloseFold(source, reportErr), &progress)
				}
				if err2 := progress.markUntraversable(true); err2 != nil {
					return nil, failureOf(sourceCloseFold(source, err2), &progress)
				}
				if err2 := source.Verify(); err2 != nil {
					return nil, failureOf(sourceCloseFold(source, err2), &progress)
				}
				if closeErr := source.Close(); closeErr != nil {
					return nil, failureOf(closeErr, &progress)
				}
				return &ValidationResult{Valid: false, FileIdentity: source.PublicIdentity(), Generation: nil, Progress: progress}, nil
			}
		}
		progress := NewProgress()
		return nil, failureOf(sourceCloseFold(source, err), &progress)
	}

	// require_available is the recorded POSIX no-op (Rust
	// live_cleanup::require_main_available); the Windows GC custody
	// arrives with the M5 surface.

	whole, err := mapping.MapFile(source.FileHandle(), res.Meta.PageCount*format.PageSize, false)
	if err != nil {
		progress := NewProgress()
		return nil, failureOf(sourceCloseFold(source, err), &progress)
	}
	defer func() { _ = whole.Close() }()
	// Rust validation.rs:310 validation_mapping applies the worker
	// session unreadable-page list to the whole-source mapping: a
	// declared page refuses deterministically with the io-unreadable
	// class instead of faulting, so a restarted worker session never
	// SIGBUSes on the same page.
	if err := whole.SetUnreadablePages(mapping.SessionUnreadablePages()); err != nil {
		progress := NewProgress()
		return nil, failureOf(sourceCloseFold(source, err), &progress)
	}

	ctx, err := newContext(whole, res.Meta, budget, check, sink)
	if err != nil {
		progress := NewProgress()
		return nil, failureOf(sourceCloseFold(source, err), &progress)
	}
	// Rust validation.rs:283 probe_source: the whole-source mapping is
	// armed with the Source role for the reserve+selected sweep, so a
	// real mapped fault lands in the worker fault record instead of
	// chaining (no hook in library processes: the Probe runs the
	// sweep directly).
	scanErr := whole.Probe(mapping.RoleSource, func() error {
		if err := ctx.reserveAllocatorPages(); err != nil {
			return err
		}
		return validateSelected(ctx)
	})
	verification := source.Verify()
	progress := ctx.finish()
	if scanErr != nil {
		return nil, failureOf(sourceCloseFold(source, scanErr), &progress)
	}
	if verification != nil {
		return nil, failureOf(sourceCloseFold(source, verification), &progress)
	}
	if closeErr := source.Close(); closeErr != nil {
		return nil, failureOf(closeErr, &progress)
	}
	return &ValidationResult{
		Valid:        progress.FindingCount == 0,
		FileIdentity: source.PublicIdentity(),
		Generation:   generation(res.Meta),
		Progress:     progress,
	}, nil
}

// validateSelected runs the structure validators over the context in the
// Rust validate_selected order (validation.rs:460) and ends with the
// allocation-partition sweep. Slice B ships the metadata and retirement
// validators plus the tree/page authorities; the remaining validators
// are the slice C-E stubs. The sweep itself runs reserveAllocatorPages
// before validateSelected inside probe_source, exactly like Rust.
func validateSelected(ctx *context) error {
	if err := validateRange(ctx); err != nil {
		return err
	}
	if err := validateCatalog(ctx); err != nil {
		return err
	}
	if err := validateStructure(ctx); err != nil {
		return err
	}
	if err := validateMembership(ctx); err != nil {
		return err
	}
	if err := validateMetadata(ctx); err != nil {
		return err
	}
	if err := validateFreeBitmap(ctx); err != nil {
		return err
	}
	if err := validateRetirement(ctx); err != nil {
		return err
	}
	return ctx.validatePartition()
}

// SweepSelected runs the committed-extent sweep of one selected
// generation over an opened source file (Rust validate_offline's
// validation_mapping + reserve_allocator_pages + validate_selected +
// finish): the generation is mapped read-only, the context is built,
// the allocator reserve and the selected graph are swept, and the
// partial progress is returned with the first failure. The caller
// owns the file, the recovery identity checks, and the terminal
// verification; a mapping or context failure returns the empty
// progress exactly like the Rust failure arms.
func SweepSelected(file *os.File, meta format.Meta, budget *ValidationBudget, check func() error, sink ValidationSink) (ValidationProgress, error) {
	if meta.PageCount > ^uint64(0)/format.PageSize {
		return ValidationProgress{}, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation mapping length"}
	}
	whole, err := mapping.MapFile(file, meta.PageCount*format.PageSize, false)
	if err != nil {
		return ValidationProgress{}, err
	}
	defer func() { _ = whole.Close() }()
	// Rust validation.rs validation_mapping (validate_offline arm)
	// applies the worker session unreadable-page list to the
	// whole-source mapping before the sweep.
	if err := whole.SetUnreadablePages(mapping.SessionUnreadablePages()); err != nil {
		return ValidationProgress{}, err
	}
	ctx, err := newContext(whole, meta, budget, check, sink)
	if err != nil {
		return ValidationProgress{}, err
	}
	// Rust validation.rs:130 probe_source (validate_offline arm): the
	// sweep runs inside the armed Source probe of the whole-source
	// mapping.
	scanErr := whole.Probe(mapping.RoleSource, func() error {
		if err := ctx.reserveAllocatorPages(); err != nil {
			return err
		}
		return validateSelected(ctx)
	})
	if scanErr != nil {
		return ctx.finish(), scanErr
	}
	return ctx.finish(), nil
}

// sourceBootstrapReportable reports whether one classified bootstrap
// refusal produces the bootstrap-finding report (Rust
// validate_immutable's Error::Format arm: every Format-class
// problem).
func sourceBootstrapReportable(problem *bootstrap.ProblemError) bool {
	return problem.Format.Code == format.CodeFormatInvalid
}

// bootstrapGeometryReport streams the pre-mapping geometry finding
// report (the short and unaligned arms of the Rust bootstrap report):
// findings first, then the untraversable mark, then the source
// verification, in the Rust bootstrap_report order.
func bootstrapGeometryReport(source *ImmutableSource, sink ValidationSink, problem *bootstrap.ProblemError, progress *ValidationProgress) (*ValidationResult, *ValidationFailure) {
	if err := writeBootstrapFindings(problem, sink, progress); err != nil {
		return nil, failureOf(sourceCloseFold(source, err), progress)
	}
	if err := progress.markUntraversable(true); err != nil {
		return nil, failureOf(sourceCloseFold(source, err), progress)
	}
	if err := source.Verify(); err != nil {
		return nil, failureOf(sourceCloseFold(source, err), progress)
	}
	if closeErr := source.Close(); closeErr != nil {
		return nil, failureOf(closeErr, progress)
	}
	return &ValidationResult{Valid: false, FileIdentity: source.PublicIdentity(), Generation: nil, Progress: *progress}, nil
}

// writeBootstrapFindings streams the bootstrap-refusal findings (Rust
// write_bootstrap_findings + report_meta_problem + report_bootstrap_
// finding): the per-page problems split on the magic class, the
// geometry classes share the FileGeometry finding, and the
// selection-class failures map to the MetaInvalid finding.
func writeBootstrapFindings(problem *bootstrap.ProblemError, sink ValidationSink, progress *ValidationProgress) error {
	switch problem.Kind {
	case bootstrap.ProblemNoBootstrapMeta:
		if err := reportMetaPage(problem.Meta0MagicInvalid, 0, sink, progress); err != nil {
			return err
		}
		return reportMetaPage(problem.Meta1MagicInvalid, 1, sink, progress)
	case bootstrap.ProblemStaticIdentityMismatch:
		return reportBootstrapFinding(ReasonMetaStaticMismatch, ObjectMeta, nil, sink, progress)
	case bootstrap.ProblemFileTooShort, bootstrap.ProblemFileUnaligned, bootstrap.ProblemHostAddressability, bootstrap.ProblemImmutableLengthMismatch:
		return reportBootstrapFinding(ReasonFileGeometryInvalid, ObjectFileGeometry, nil, sink, progress)
	default:
		return reportBootstrapFinding(ReasonMetaInvalid, ObjectMeta, nil, sink, progress)
	}
}

// reportMetaPage reports one per-page meta problem (Rust
// report_meta_problem): the magic class selects the MetaUnavailable
// reason, everything else is MetaInvalid, both on the Meta object
// with the page number and its physical byte interval.
func reportMetaPage(magicInvalid bool, pageNumber uint32, sink ValidationSink, progress *ValidationProgress) error {
	reason := ReasonMetaInvalid
	if magicInvalid {
		reason = ReasonMetaUnavailable
	}
	return reportBootstrapFinding(reason, ObjectMeta, &pageNumber, sink, progress)
}

// reportBootstrapFinding counts and streams one bootstrap finding
// (Rust report_bootstrap_finding: the sequence is the post-count
// finding count, and a page-carrying finding always includes its
// physical byte interval).
func reportBootstrapFinding(reason ValidationReason, object ValidationObject, pageNumber *uint32, sink ValidationSink, progress *ValidationProgress) error {
	var interval *PhysicalByteInterval
	if pageNumber != nil {
		value, err := partitionBytes(uint64(*pageNumber), uint64(*pageNumber)+1)
		if err != nil {
			return err
		}
		interval = &value
	}
	return emitFinding(progress, sink, reason, object, pageNumber, interval, nil)
}

// sourceFileSize is the physical extent of the opened source (Rust
// open_read_only fstat; the bootstrap selection requires the exact
// committed length for the immutable reader).
func sourceFileSize(source *ImmutableSource) uint64 {
	size, err := source.FileHandle().Stat()
	if err != nil {
		return 0
	}
	return uint64(size.Size())
}

// sourceCloseFold closes the source and returns the primary error
// (the close error surfaces only when the primary is nil, like the
// Rust combine_errors arms).
func sourceCloseFold(source *ImmutableSource, primary error) error {
	closeErr := source.Close()
	if primary != nil {
		return primary
	}
	return closeErr
}

// failureOf builds one operational failure with the partial progress
// and the clean ledger (Rust validation::failure).
func failureOf(cause error, progress *ValidationProgress) *ValidationFailure {
	return &ValidationFailure{
		Cause:               cause,
		Progress:            progress,
		CoordinationCleanup: publication.CoordinationCleanupNone,
	}
}
